### PR TITLE
Add Action graph API

### PR DESCRIPTION
**Pull request by jacobperron**
Connects to https://github.com/ros2/ros2cli/issues/202

### CODE REVIEW

**Comment by jacobperron** [comment]
CI up to rcl_action:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6581/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2919/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5397/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6390/)

**Comment by jacobperron** [comment]
I forgot this was based on top of https://github.com/ros2/rcl/pull/404, which should be reviewed/merged first.

**Comment by jacobperron** [comment]
Ready for review.

File: rcl_action/src/rcl_action/graph.c [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      // Populate types list
+      for (size_t k = 0u; k < topic_names_and_types->types[i].size; ++k) {
+        char * type_name = topic_names_and_types->types[i].data[k];
+        size_t action_type_len = strlen(type_name);
+        // Trim type name suffix, if provided
+        if (action_type_suffix) {
```
Comment (sloretz) [inline rcl_action/src/rcl_action/graph.c]: This is always true, unless this variable is shadowed and I missed it

```
 const char * action_type_suffix = "_FeedbackMessage";
```

File: rcl_action/src/rcl_action/graph.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  const char * node_name,
+  const char * node_namespace,
+  rcl_names_and_types_t * action_names_and_types)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_names_and_types, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret;
+  rcl_names_and_types_t topic_names_and_types = rcl_get_zero_initialized_names_and_types();
+  ret = rcl_get_subscriber_names_and_types_by_node(
+    node, allocator, false, node_name, node_namespace, &topic_names_and_types);
```
Comment (sloretz) [inline rcl_action/src/rcl_action/graph.c]: Does `topic_names_and_types` get `fini`'d anywhere? Same question in the other two methods below

File: rcl_action/CMakeLists.txt [hunk 3]
```diff
@@ -186,6 +199,19 @@ if(BUILD_TESTING)
     )
     ament_target_dependencies(test_action_server "rcl")
   endif()
+  ament_add_gtest(test_graph
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/CMakeLists.txt]: You're right. Removed f9c7161

File: rcl_action/src/rcl_action/graph.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  rcl_ret_t ret;
+  const size_t num_names = topic_names_and_types->names.size;
+  char ** names = topic_names_and_types->names.data;
+
+  // Count number of actions to determine how much memory to allocate
+  size_t num_actions = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * suffix_index = strstr(names[i], action_name_suffix);
+    if (suffix_index) {
+      num_actions++;
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Fixed e738aaa

File: rcl_action/src/rcl_action/graph.c [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        }
+        char * action_type_name = rcutils_strndup(type_name, action_type_len, *allocator);
+        if (!action_type_name) {
+          RCL_SET_ERROR_MSG("Failed to allocate memory for action type");
+          ret = RCL_RET_BAD_ALLOC;
+          break;
+        }
+        action_names_and_types->types[j].data[k] = action_type_name;
+      }
+      j++;
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Fixed e738aaa

File: rcl_action/src/rcl_action/graph.c [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  const char * action_type_suffix = "_FeedbackMessage";
+
+  rcl_ret_t ret;
+  const size_t num_names = topic_names_and_types->names.size;
+  char ** names = topic_names_and_types->names.data;
+
+  // Count number of actions to determine how much memory to allocate
+  size_t num_actions = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * suffix_index = strstr(names[i], action_name_suffix);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Renamed to 'identifier' 91c10ca

File: rcl_action/src/rcl_action/graph.c (hunk 1 above)
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Yep. Removed 2923012

File: rcl_action/src/rcl_action/graph.c (hunk 2 above)
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Good catch! Fixed 2923012

File: rcl_action/src/rcl_action/graph.c [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  ret = _filter_action_names(
+    &topic_names_and_types,
+    allocator,
+    action_names_and_types);
+
+  rcl_ret_t nat_fini_ret = rcl_names_and_types_fini(&topic_names_and_types);
+  if (RCL_RET_OK != nat_fini_ret) {
+    return nat_fini_ret;
```
Comment (sloretz) [inline rcl_action/src/rcl_action/graph.c]: Recommend `fini`'ing `action_names_and_types` here so the user doesn't have to do it when the function returns an error code, same below.

File: rcl_action/src/rcl_action/graph.c [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    (void)fini_ret;  // Error already set
+  }
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_action_get_client_names_and_types_by_node(
+  const rcl_node_t * node,
+  rcl_allocator_t * allocator,
```
Comment (sloretz) [inline rcl_action/src/rcl_action/graph.c]: I don't see anything checking if the allocator is valid.

may need to use `RCL_CHECK_ALLOCATOR`

https://github.com/ros2/rcl/blob/d810f12e94ae9498f34ebcf77951b544d21aa10f/rcl/include/rcl/allocator.h#L31-L32

and add a test with a zero initialized allocator

https://github.com/ros2/rcutils/blob/1cb9d67ea7e3ca848d7477e381257776ac00ba85/include/rcutils/allocator.h#L83-L84

File: rcl_action/test/rcl_action/test_graph.cpp [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+
+/**
+ * Extend the TestActionGraphFixture with a multi-node fixture for node discovery and node-graph
+ * perspective.
+ */
+class TestActionGraphMultiNodeFixture : public CLASSNAME(TestActionGraphFixture, RMW_IMPLEMENTATION)
+{
+public:
+  const char * remote_node_name = "remote_graph_node";
+  const char * action_name = "/test_action_info_functions__";
```
Comment (sloretz) [inline rcl_action/test/rcl_action/test_graph.cpp]: http://design.ros2.org/articles/topic_and_service_names.html forbids `__`

> * must not contain any number of repeated underscores (`_`)

I'm surprised the action topics and services are created successfully. This might be worth opening a bug for.

Comment (jacobperron) [reply rcl_action/test/rcl_action/test_graph.cpp]: I didn't realize, I copied this from what's done in the `rcl` graph tests:

https://github.com/ros2/rcl/blob/d810f12e94ae9498f34ebcf77951b544d21aa10f/rcl/test/rcl/test_graph.cpp#L737

File: rcl_action/test/rcl_action/test_graph.cpp [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+    &this->node, nullptr, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid node name
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, &this->allocator, "_test_this_Isnot_a_valid_name", "", nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, &this->allocator, this->test_graph_old_node_name, "", nullptr);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/test/rcl_action/test_graph.cpp]: It's invalid because the node has been shutdown and is no longer valid. Added a separate comment for this case 57e2280

File: rcl_action/test/rcl_action/test_graph.cpp [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+  rcl_reset_error();
+  // Invalid allocator
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, nullptr, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid node name
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, &this->allocator, "_test_this_Isnot_a_valid_name", "", nullptr);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/test/rcl_action/test_graph.cpp]: After passing `&nat` here, I released that it actually returns `RCL_RET_ERROR`. This because of the underlying call to `rcl_get_subscriber_names_and_types_by_node`, which has the same behavior.

File: rcl_action/test/rcl_action/test_graph.cpp [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid allocator
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, nullptr, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid node name
+  ret = rcl_action_get_client_names_and_types_by_node(
+    &this->node, &this->allocator, "_test_this_Isnot_a_valid_name", "", nullptr);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/test/rcl_action/test_graph.cpp]: Done a70eac7

File: rcl_action/test/rcl_action/test_graph.cpp [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid allocator
+  ret = rcl_action_get_server_names_and_types_by_node(
+    &this->node, nullptr, this->test_graph_node_name, "", &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // Invalid node name
+  ret = rcl_action_get_server_names_and_types_by_node(
+    &this->node, &this->allocator, "_test_this_Isnot_a_valid_name", "", nullptr);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/test/rcl_action/test_graph.cpp]: a70eac7

File: rcl_action/test/rcl_action/test_graph.cpp [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_action_get_server_names_and_types_by_node(
+    &this->node, &this->allocator, this->test_graph_old_node_name, "", nullptr);
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl_action/test/rcl_action/test_graph.cpp]: a70eac7

File: rcl_action/src/rcl_action/graph.c (hunk 8 above)
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Added tests 3adab6b

File: rcl_action/src/rcl_action/graph.c (hunk 7 above)
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Done 57e2280

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6670/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2989/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5478/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6504/)

File: rcl_action/CMakeLists.txt [hunk 15]
```diff
@@ -146,7 +152,14 @@ if(BUILD_TESTING)
...
       "test/rcl_action/test_action_interaction.cpp")
-  endmacro()
+
+    # TODO(jacobperron): Graph tests fail with opensplice. Re-enable after resolving
+    # https://github.com/ros2/ros2/issues/677
+    if(NOT rmw_implementation STREQUAL "rmw_opensplice_cpp")
+      custom_test_c(test_graph
+        "test/rcl_action/test_graph.cpp")
+    endif()
+endmacro()
```
Comment (dirk-thomas) [inline rcl_action/CMakeLists.txt]: Indentations seems to be off.

File: rcl_action/src/rcl_action/graph.c [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+  const char * action_type_identifier = "_FeedbackMessage";
+
+  rcl_ret_t ret;
+  const size_t num_names = topic_names_and_types->names.size;
+  char ** names = topic_names_and_types->names.data;
+
+  // Count number of actions to determine how much memory to allocate
+  size_t num_actions = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * identifier_index = strstr(names[i], action_name_identifier);
```
Comment (dirk-thomas) [inline rcl_action/src/rcl_action/graph.c]: Should this actually check if `action_name_identifier` is a suffix and not anywhere in the middle of the string as the command above suggests?

File: rcl_action/CMakeLists.txt (hunk 15 above)
Comment (jacobperron) [reply rcl_action/CMakeLists.txt]: Fixed 9e9c238

File: rcl_action/src/rcl_action/graph.c (hunk 16 above)
Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Yes, that would be stronger, but based on this comment, https://github.com/ros2/rcl/pull/411#discussion_r274555643, I thought it sufficient to leave this as-is. But, if you have a recommendation for a nice way to ensure it is at the end of the string I'm happy to add it.

File: rcl_action/src/rcl_action/graph.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (dirk-thomas) [reply rcl_action/src/rcl_action/graph.c]: This falls under the category of preferring defensive programming. If you can make it more concrete / specific then better do so - either it might hit you back. So I think checking the condition to actually be a suffix would be better - and it is literally just one more line.

File: rcl_action/src/rcl_action/graph.c (hunk 16 above)
Comment (dirk-thomas) [reply rcl_action/src/rcl_action/graph.c]: > if you have a recommendation for a nice way to ensure it is at the end of the string I'm happy to add it.

`identifier_index` needs to be `strlen(names[i]) - strlen(action_name_identifier)`?

Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: Since `strstr` returns a reference to the start of the substring, I think we want `strlen(identifier_index) == strlen(action_name_identifier)`, but seems like it's broken some tests. Looking into it.

Comment (jacobperron) [reply rcl_action/src/rcl_action/graph.c]: I think this does it: 661cab0

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6681/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3000/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5489/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6511/)
